src/data/loaders.py

```diff
@@ -1,113 +1,423 @@
+        new_active_pitch_notes = [(pitch, end) for pitch, end
+            in active_pitch_notes[program_family] if end > quantized_start]
+        active_pitch_notes[program_family] = new_active_pitch_notes
+        note_pitch = midi_note.pitch
+        max_end = 0
+        for pitch, end in active_pitch_notes[program_family]:
+            if pitch == note_pitch and end > max_end:
+                max_end = end
+        if max_end >= quantized_end:
+            continue
+        quantized_start = max(quantized_start, max_end)
+        active_pitch_notes[program_family].append((note_pitch, quantized_end))
+        no_clash_notes.append((quantized_start, quantized_end, program, midi_note))
 
-    def __call__(self, filepath):
-        """This method takes a MIDI file path and returns a list of integer
-        event IDs.
-        """
-        raise NotImplementedError
+    return no_clash_notes
+
+
+def get_event_list(midi_notes, num_velocity_bins=32):
+    no_drum_notes = [(quantized_start, quantized_end, program, midi_note)
+        for program, instrument, is_drum, quantized_start, quantized_end, midi_note
+            in midi_notes if not is_drum]
+    sorted_no_drum_notes = sorted(no_drum_notes, key=lambda element: element[0:3])
+    no_clash_notes = resolve_pitch_clashes(sorted_no_drum_notes)
+
+    note_on_set = []
+    note_off_set = []
+    for index, element in enumerate(no_clash_notes):
+        quantized_start, quantized_end, program, midi_note = element
+        note_on_set.append((quantized_start, index, program, False))
+        note_off_set.append((quantized_end, index, program, True))
+    note_events = sorted(note_on_set + note_off_set)
+
+    velocity_bin_size = int(math.ceil(
+        (MAX_MIDI_VELOCITY - MIN_MIDI_VELOCITY + 1) / num_velocity_bins))
+    num_program_families = int((MAX_MIDI_PROGRAM - MIN_MIDI_PROGRAM + 1) / \
+        PROGRAMS_PER_FAMILY)
+
+    current_step = 0
+    current_velocity_bin = {}
+    for program_family in range(num_program_families):
+        current_velocity_bin[program_family + 1] = 0
+    events = []
+
+    for step, index, program, is_off in note_events:
+        if step > current_step:
+            while step > current_step + MAX_SHIFT_STEPS:
+                events.append((TIME_SHIFT, MAX_SHIFT_STEPS, 0))
+                current_step += MAX_SHIFT_STEPS
+            events.append((TIME_SHIFT, step-current_step, 0))
+            current_step = step
+
+        note_velocity = no_clash_notes[index][3].velocity
+        note_pitch = no_clash_notes[index][3].pitch
+        velocity_bin = (note_velocity - MIN_MIDI_VELOCITY) // velocity_bin_size + 1
+        program_family = (program - MIN_MIDI_PROGRAM) // PROGRAMS_PER_FAMILY + 1
+        if not is_off and velocity_bin != current_velocity_bin[program_family]:
+            current_velocity_bin[program_family] = velocity_bin
+            events.append((VELOCITY, velocity_bin, program_family))
```

Comment:
Can we add a note that velocity bin is currently set per program family and a future TODO comment of possibly having the velocity bin be independent of the program family as this would decrease the space of the output at each step?

Reply:
Will do!